utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/encode/AvroSchemaOutputValidator.scala

```diff
@@ -0,0 +1,285 @@
+    }
+
+   requiredFieldsValidation combine schemaFieldsValidation combine redundantFieldsValidation
+  }
+
+  private def validateMapSchema(typingResult: TypingResult, schema: Schema, path: Option[String]) = {
+    def isMap(klass: Class[_]) = classOf[java.util.Map[_, _]].isAssignableFrom(klass)
+
+    typingResult match {
+      case _@TypedClass(klass, key :: value :: Nil) if isMap(klass) =>
+        //Map keys are assumed to be strings: https://avro.apache.org/docs/current/spec.html#Maps
+        condNel(key.canBeSubclassOf(Typed.apply[java.lang.String]), (), typeError(typingResult, schema, path)).andThen(_ =>
+          validateTypingResult(value, schema.getValueType, buildPath("*", path, useIndexer = true))
+        )
+      case map@TypedClass(klass, _) if isMap(klass) =>
+        throw new IllegalArgumentException(s"Illegal typing Map: $map.")
+      case _@TypedObjectTypingResult(fields, _, _) =>
+        fields.map{ case (key, value) =>
+          val fieldPath = buildPath(key, path)
+          validateTypingResult(value, schema.getValueType, fieldPath)
+        }.foldLeft[ValidatedNel[OutputValidatorError, Unit]](().validNel)((a, b) => a combine b)
+      case _ =>
+        invalid(typingResult, schema, path)
+    }
+  }
+
+  private def validateArraySchema(typedClass: TypedClass, schema: Schema, path: Option[String]) =
+    typedClass.params match {
+      case head :: Nil =>
+        val elementPath = buildPath("", path, useIndexer = true)
+        validateTypingResult(head, schema.getElementType, elementPath)
+      case _ =>
+        throw new IllegalArgumentException(s"Illegal typing List: $typedClass.")
+    }
+
+  private def validateUnionSchema(typingResult: TypingResult, schema: Schema, path: Option[String]) = {
+    //check is there only one typing error with exactly same field as path - it means there was checking whole object (without going deeper e.g. List/Map/Record)
+    def singleObjectTypingError(errors: NonEmptyList[OutputValidatorError]): Boolean =
+      errors.collect{case err: OutputValidatorTypeError => err} match {
+        case head :: Nil => path.contains(head.field)
+        case _ => false
+      }
+
+    def createUnionValidationResults(checkNullability: Boolean): List[ValidatedNel[OutputValidatorError, Unit]] = {
+      val schemas = if(checkNullability) schema.getTypes.asScala else schema.getTypes.asScala.filterNot(_.isNullable)
+      schemas.map(validateTypingResult(typingResult, _, path)).toList
+    }
+
+    def isNullableSchema(schema: Schema): Boolean = schema.getTypes.size() == 2 && schema.isNullable
+
+    def asSingleValidatedResults(results: List[ValidatedNel[OutputValidatorError, Unit]]) =
+      if (results.exists(_.isValid)) valid else results.sequence.map(_=> ())
+
+    val unionValidationResults = schema match {
+      case sch if isNullableSchema(sch) =>
```

Comment:
It is not exactly nullable schema. To be precise it is union of null and one more type. I wonder if won't be better to just inline this expression here and add a little comment that we handle it that way to not overwhelm messages with X is not null

Reply:
Inlined.